### DISCUSSION TITLE
Inconsistency in DNS stamps spec vs calculator vs usage in dnscrypt-proxy

### DISCUSSION THREAD

**Discussion by cobratbq** [DNSCrypt/dnscrypt-proxy#1862] [Potential issues]
In follow-up of #1861, I looked into something that confused me yesterday-night: there seems to be an inconsistency in how DNS stamps are defined vs. implemented.

## Specification, section "DNS-over-HTTPS stamps"

> addr is the IP address of the server. It can be an empty string, or just a port number, represented with a preceding colon (:443). In that case, the host name will be resolved to an IP address using another resolver.

Implying that the port number could optionally be added here.

> hostname is the server host name which will also be used as a SNI name. If the host name contains characters outside the URL-permitted range, these characters should be sent as-is, without any extra encoding (neither URL-encoded nor punycode).

No mention of port number at all.

## Online DNS Stamp calculator

The following labels as shown in the calculator ...

> IP Address (IPv6 addresses must be in [ ] brackets)

No mention of port number.

> Host name (vhost + SNI) and optional port number

Explicitly mentioned port number.

## Inconsistency

The implementation (_dnscrypt-proxy_) seems to follow the spec primarily, i.e. not expecting the port in the host name (`ProviderName`) field. I think either calculator or spec may need clearing up.
Somewhat counterintuitively (maybe), it makes sense to have the "server address" only be an address. This makes it seem more like the optionally added bootstrapping information, while the host name contains the full location, i.e. with port number.

@jedisct1 I just want to point this out to make sure you are aware of it.

**Comment by jedisct1** [comment]
Thanks! This has been clarified in the specification.

Having the port number in the address as originally specified is would actually make more sense, and be consistent with DNSCrypt. However, current implementations expect it to be after the hostname, so the specification should reflect what's actually deployed.

Thanks!
